### PR TITLE
add support for pushsafer.com notification service

### PR DESCRIPTION
**Pull request by appzer**
API description: https://www.pushsafer.com/en/pushapi

==== Pushsafer ====

**Version**\\
Available since pilight v7.0

**Description**\\
Send a message to your iOS, Android or Windows 10 (Phone, Dekstop) device with the Pushsafer API

**Options**
^name^required^multiple values^description^
|TITLE|x| |Self-explanatory|
|MESSAGE|x| |Self-explanatory|
|PRIVATEKEY|x| |Unique private or alias key from the pushsafer website|
|DEVICE|| |Unique device or device group id from the pushsafer website|
|ICON|| |a number 1-98 as described in the API|
|SOUND|| |a number 0-28 or empty as described in the API|
|VIBRATION|| |a number 1-3 or empty as described in the API|

**Examples**
<code>
IF 1 == 1 THEN pushsafer TITLE Doorbell rang MESSAGE Doorbell rang PRIVATEKEY abcd123abc
</code>

<code>
IF 1 == 1 THEN pushsafer TITLE Doorbell rang MESSAGE Doorbell rang PRIVATEKEY abcd123abc DEVICE 101 ICON 5 SOUND 11 VIBRATION 2
</code>

### CODE REVIEW

**Comment by appzer** [comment]
Hi CurlyMoo,

when you will merge this pr.

Kevin

**Comment by CurlyMoo** [comment]
When i have time to look into it.

File: libs/pilight/events/actions/pushsafer.c [hunk 1]
```diff
@@ -0,0 +1,311 @@
+/*
+	Copyright (C) 2013 - 2014 CurlyMo
+	
+	Copyright (C) 2016 Pushsafer.com Kevin Siml > Appzer.de
```
Comment (CurlyMoo) [inline libs/pilight/events/actions/pushsafer.c]: Can you make the Copyright one single line with your info (and mine if you like).

File: libs/pilight/events/actions/pushsafer.c [hunk 2]
```diff
@@ -0,0 +1,311 @@
...
+
+	This file is part of pilight.
```
Comment (CurlyMoo) [inline libs/pilight/events/actions/pushsafer.c]: I wan't new code to be submitted under the MPLv2 license.

File: libs/pilight/events/actions/pushsafer.c [hunk 3]
```diff
@@ -0,0 +1,311 @@
...
+
+	jtitle = json_find_member(obj->arguments, "TITLE");
+	jmessage = json_find_member(obj->arguments, "MESSAGE");
+	jprivatekey = json_find_member(obj->arguments, "PRIVATEKEY");
+	jdevice = json_find_member(obj->arguments, "DEVICE");
+	jicon = json_find_member(obj->arguments, "ICON");
+	jsound = json_find_member(obj->arguments, "SOUND");
+	jvibration = json_find_member(obj->arguments, "VIBRATION");
+
+	if(jtitle == NULL) {
```
Comment (CurlyMoo) [inline libs/pilight/events/actions/pushsafer.c]: Are all these arguments required?

File: libs/pilight/events/actions/pushsafer.c [hunk 4]
```diff
@@ -0,0 +1,311 @@
...
+	if(nrvalues != 1) {
+		logprintf(LOG_ERR, "pushsafer action \"VIBRATION\" only takes one argument");
+		return -1;
+	}	
+	return 0;
+}
+
+static void *thread(void *param) {
+	struct rules_actions_t *pth = (struct rules_actions_t *)param;
+	// struct rules_t *obj = pth->obj;
```
Comment (CurlyMoo) [inline libs/pilight/events/actions/pushsafer.c]: If you are not using this, then just remove it.

File: libs/pilight/events/actions/pushsafer.c [hunk 5]
```diff
@@ -0,0 +1,311 @@
...
+
+	jtitle = json_find_member(arguments, "TITLE");
+	jmessage = json_find_member(arguments, "MESSAGE");
+	jprivatekey = json_find_member(arguments, "PRIVATEKEY");
+	jdevice = json_find_member(arguments, "DEVICE");
+	jicon = json_find_member(arguments, "ICON");
+	jsound = json_find_member(arguments, "SOUND");
+	jvibration = json_find_member(arguments, "VIBRATION");
+
+	if(jtitle != NULL && jmessage != NULL && jprivatekey != NULL && jdevice != NULL && jicon != NULL && jsound != NULL && jvibration != NULL) {
```
Comment (CurlyMoo) [inline libs/pilight/events/actions/pushsafer.c]: Again, is this all required? I would also word wrap this.

File: libs/pilight/events/actions/pushsafer.c [hunk 6]
```diff
@@ -0,0 +1,311 @@
...
+		jvalues1 = json_find_member(jtitle, "value");
+		jvalues2 = json_find_member(jmessage, "value");
+		jvalues3 = json_find_member(jprivatekey, "value");
+		jvalues4 = json_find_member(jdevice, "value");
+		jvalues5 = json_find_member(jicon, "value");
+		jvalues6 = json_find_member(jsound, "value");
+		jvalues7 = json_find_member(jvibration, "value");
+		if(jvalues1 != NULL && jvalues2 != NULL && jvalues3 != NULL && jvalues4 != NULL && jvalues5 != NULL && jvalues6 != NULL && jvalues7 != NULL) {
```
Comment (CurlyMoo) [inline libs/pilight/events/actions/pushsafer.c]: Likewise.

File: libs/pilight/events/actions/pushsafer.c [hunk 7]
```diff
@@ -0,0 +1,311 @@
...
+			jval1 = json_find_element(jvalues1, 0);
+			jval2 = json_find_element(jvalues2, 0);
+			jval3 = json_find_element(jvalues3, 0);
+			jval4 = json_find_element(jvalues4, 0);
+			jval5 = json_find_element(jvalues5, 0);
+			jval6 = json_find_element(jvalues6, 0);
+			jval7 = json_find_element(jvalues7, 0);
+			if(jval1 != NULL && jval2 != NULL && jval3 != NULL && jval4 != NULL && jval5 != NULL && jval6 != NULL && jval7 != NULL &&
```
Comment (CurlyMoo) [inline libs/pilight/events/actions/pushsafer.c]: Likewise

File: libs/pilight/events/actions/pushsafer.c [hunk 8]
```diff
@@ -0,0 +1,311 @@
...
+			 jval1->tag == JSON_STRING && jval2->tag == JSON_STRING &&
+			 jval3->tag == JSON_STRING && jval4->tag == JSON_STRING &&
+			 jval5->tag == JSON_STRING && jval6->tag == JSON_STRING &&
+			 jval7->tag == JSON_STRING) {
+				data = NULL;
+				strcpy(url, "https://www.pushsafer.com/api");
```
Comment (CurlyMoo) [inline libs/pilight/events/actions/pushsafer.c]: Why don't you urlencode the final sprintf?

File: libs/pilight/events/actions/pushsafer.c (hunk 3 above)
Comment (appzer) [inline libs/pilight/events/actions/pushsafer.c]: only message and privatekey are required, all other are optional

File: libs/pilight/events/actions/pushsafer.c (hunk 5 above)
Comment (appzer) [inline libs/pilight/events/actions/pushsafer.c]: only message and privatekey are required, all other are optional

File: libs/pilight/events/actions/pushsafer.c (hunk 3 above)
Comment (CurlyMoo) [inline libs/pilight/events/actions/pushsafer.c]: But if any of them are zero you give an error.

Comment (appzer) [inline libs/pilight/events/actions/pushsafer.c]: i know, but i think the user should set up the notification fully

Comment (CurlyMoo) [inline libs/pilight/events/actions/pushsafer.c]: Thus they are required?

Comment (appzer) [inline libs/pilight/events/actions/pushsafer.c]: yes

Comment (appzer) [inline libs/pilight/events/actions/pushsafer.c]: now everything is OK?

Comment (CurlyMoo) [inline libs/pilight/events/actions/pushsafer.c]: > only message and privatekey are required, all other are optional

If this is true, then i would make them optional.

Comment (appzer) [inline libs/pilight/events/actions/pushsafer.c]: its true, can you make the changes please

Comment (CurlyMoo) [inline libs/pilight/events/actions/pushsafer.c]: No, i can't influence your PR.

Comment (appzer) [inline libs/pilight/events/actions/pushsafer.c]: i updated the file, hope all is correct

File: libs/pilight/events/actions/pushsafer.c [hunk 9]
```diff
@@ -0,0 +1,225 @@
...
+	int nrvalues = 0;
+
+	jtitle = json_find_member(obj->arguments, "TITLE");
+	jmessage = json_find_member(obj->arguments, "MESSAGE");
+	jprivatekey = json_find_member(obj->arguments, "PRIVATEKEY");
+	jdevice = json_find_member(obj->arguments, "DEVICE");
+	jicon = json_find_member(obj->arguments, "ICON");
+	jsound = json_find_member(obj->arguments, "SOUND");
+	jvibration = json_find_member(obj->arguments, "VIBRATION");
+
```
Comment (CurlyMoo) [inline libs/pilight/events/actions/pushsafer.c]: If users do set these optional values, you still need to validate them.

File: libs/pilight/events/actions/pushsafer.c [hunk 10]
```diff
@@ -0,0 +1,225 @@
...
+	jtitle = json_find_member(arguments, "TITLE");
+	jmessage = json_find_member(arguments, "MESSAGE");
+	jprivatekey = json_find_member(arguments, "PRIVATEKEY");
+	jdevice = json_find_member(arguments, "DEVICE");
+	jicon = json_find_member(arguments, "ICON");
+	jsound = json_find_member(arguments, "SOUND");
+	jvibration = json_find_member(arguments, "VIBRATION");
+
+	if(jmessage != NULL && jprivatekey != NULL) {
+		jvalues1 = json_find_member(jtitle, "value");
```
Comment (CurlyMoo) [inline libs/pilight/events/actions/pushsafer.c]: When values aren't set, this function will probably fail.

File: libs/pilight/events/actions/pushsafer.c [hunk 11]
```diff
@@ -0,0 +1,225 @@
...
+				data = NULL;
+				strcpy(url, "https://www.pushsafer.com/api");
+				char *message = jval2->string_;
+				char *privatekey = jval3->string_;
+				char *device = jval4->string_;
+				char *icon = jval5->string_;
+				char *sound = jval6->string_;
+				char *vibration = jval7->string_;
+				char *title = jval1->string_;
+				size_t l = strlen(message)+strlen(privatekey);
```
Comment (CurlyMoo) [inline libs/pilight/events/actions/pushsafer.c]: Again, what happens if users didn't set these values.

File: libs/pilight/events/actions/pushsafer.c [hunk 12]
```diff
@@ -0,0 +1,225 @@
...
+				l += strlen(device)+strlen(title);
+				l += strlen(icon)+strlen(sound)+strlen(vibration);
+				l += strlen("k=")+strlen("&d=");
+				l += strlen("&m=")+strlen("&t=");
+				l += strlen("&i=")+strlen("&s=")+strlen("&v=");
+				char content[l+2];
+				sprintf(content, "k=%s&d=%s&i=%s&s=%s&v=%s&t=%s&m=%s", privatekey, device, icon, sound, vibration, title, message);
```
Comment (CurlyMoo) [inline libs/pilight/events/actions/pushsafer.c]: This should adapt itself to the values set.

File: libs/pilight/events/actions/pushsafer.h [hunk 13]
```diff
@@ -0,0 +1,30 @@
+/*
+	Copyright (C) 2013 - 2014 CurlyMo
+	
+	Copyright (C) 2016 Pushsafer.com Kevin Siml > Appzer.de
```
Comment (CurlyMoo) [inline libs/pilight/events/actions/pushsafer.h]: You forgot these.

Comment (CurlyMoo) [inline libs/pilight/events/actions/pushsafer.h]: And ofc the license.

File: libs/pilight/events/actions/pushsafer.c (hunk 9 above)
Comment (appzer) [inline libs/pilight/events/actions/pushsafer.c]: Ok, i made my changes

File: libs/pilight/events/actions/pushsafer.c (hunk 10 above)
Comment (appzer) [inline libs/pilight/events/actions/pushsafer.c]: OK i made my changes, is all ok now

Is this part OK?
jval1->tag == JSON_STRING && jval2->tag == JSON_STRING &&
			 jval3->tag == JSON_STRING && jval4->tag == JSON_STRING &&
			 jval5->tag == JSON_STRING && jval6->tag == JSON_STRING &&
			 jval7->tag == JSON_STRING

**Comment by appzer** [comment]
CurlyMoo, i made my changes. Is all OK now?

File: libs/pilight/events/actions/pushsafer.c (hunk 9 above)
Comment (CurlyMoo) [inline libs/pilight/events/actions/pushsafer.c]: I still don't see the validation for the other arguments.

File: libs/pilight/events/actions/pushsafer.c (hunk 10 above)
Comment (CurlyMoo) [inline libs/pilight/events/actions/pushsafer.c]: You still assume that all arguments are set.

Comment (appzer) [inline libs/pilight/events/actions/pushsafer.c]: made changes.
Is this a part of validation?
nrvalues = 0;
	if((jvalues = json_find_member(jmessage, "value")) != NULL) {
		jchild = json_first_child(jvalues);
		while(jchild) {
			nrvalues++;
			jchild = jchild->next;
		}
	}
	if(nrvalues != 1) {
		logprintf(LOG_ERR, "pushsafer action \"MESSAGE\" only takes one argument");
		return -1;
	}

Comment (CurlyMoo) [inline libs/pilight/events/actions/pushsafer.c]: Can you use code formatting when posting here?

Comment (appzer) [inline libs/pilight/events/actions/pushsafer.c]: done
Is this a validation you mean?

Comment (CurlyMoo) [inline libs/pilight/events/actions/pushsafer.c]: You should test your own code with all kinds of argument combination and check if it works. Also check for invalid combinations and check if you don't get errors. These tests (and their solutions) should be part of every developers workflow.

Comment (appzer) [inline libs/pilight/events/actions/pushsafer.c]: OK, now all is done

Comment (CurlyMoo) [inline libs/pilight/events/actions/pushsafer.c]: There are no new commits. Also, please post all test you've run.

Comment (CurlyMoo) [inline libs/pilight/events/actions/pushsafer.c]: Can you show me all tests you've run?

Comment (appzer) [inline libs/pilight/events/actions/pushsafer.c]: i do not have the environment to test the code. I only adapt the code from the pushover plugin to my needs. Sorry! Kevin

**Comment by CurlyMoo** [comment]
> i do not have the environment to test the code. I only adapt the code from the pushover plugin to my needs. Sorry! Kevin

Then i do consider this effort useless. Don't open new PR's but let me know when you do have a test environment and we will continue from there.

**Pull request closed without merging**